tests/PHPStan/Analyser/data/TestDynamicReturnTypeExtensions.php

```diff
@@ -240,3 +241,36 @@ public function getTypeFromMethodCall(
 	}
 
 }
+
+class Bug7391DynamicStaticMethodReturnTypeExtension implements DynamicStaticMethodReturnTypeExtension
+{
+	public function getClass(): string
+	{
+		return \Bug7391\Foo::class;
+	}
+
+	public function isStaticMethodSupported(MethodReflection $methodReflection): bool
+	{
+		return $methodReflection->getName() === 'm';
+	}
+
+	private function getMethodCallScopeType(StaticCall $methodCall, Scope $scope): Type
+	{
+		$classNameType = $scope->getType(new ClassConstFetch($methodCall->class, 'class'));
```

Comment:
@rvanvelzen please see the failing test, currently this results in "error type"

with https://github.com/phpstan/phpstan-src/pull/1416#issuecomment-1155787778 while loop unroll/prepare, it works correctly, but that should be done inside the `$scope->getType()`

Reply:
The problem is that `(X::class)::class` is not valid, so `$scope->getType()` _should_ return `ErrorType` for it

Reply:
Thank you, you are right!

I should probably not use `ClassConstFetch` to resolve the string type to object type.

But what should I use instead? `$scope->resolveTypeByName()` cannot be used as it accepts only `Name`.

With my solution, I solved at least `(self::class)::m()` as for invocations the class can be a string.

But my solution would fail for `get_class($foo)::m()`.

I am looking for some universal solution, which will resolve any class string type (generic,, unioned, ...) to object type.

Reply:
I tried:

- `return $scope->getType(new MethodCall(new New_(new String_(\ReflectionClass::class), [$methodCall->class]), 'newInstanceWithoutConstructor'));`, but it returns `object` type
- `return $scope->getType(new New_($methodCall->class));`, but it returns `*ERROR*` type